compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -4289,6 +4289,28 @@ private BType getIntersection(IntersectionContext intersectionContext, BType lhs
                 return lhsType;
             }
         }
+
+        // TODO: intersections with readonly types are not handled properly. Here, the logic works as follows.
+        // Say we have an intersection called A & readonly and we have another type called B. As per the current
+        // implementation, we cannot easily find the intersection between (A & readonly) and B. Instead, what we
+        // do here is, first find the intersection between A and B then re-create the immutable type out of it.
+
+        if (Symbols.isFlagOn(lhsType.flags, Flags.READONLY) && lhsType.tag == TypeTags.UNION &&
+                ((BUnionType) lhsType).getIntersectionType().isPresent()) {
```

Comment:
Shouldn't we check this for `type` also?

Reply:
Will address with this issue separately: https://github.com/ballerina-platform/ballerina-lang/issues/33598